### PR TITLE
[Tune] Unrevert "Add more comprehensive support for remote_checkpoint_dir w/ url params (#32479)"

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This reverts the revert made here #32571.

This PR fixes two suites of tests that were failing:
1. workflows tests were failing due to the pinned `moto` version being too old.
2. `test_syncer` errors were caused by reusing the same bucket name between tests.

This PR also fixes the `restore` path for URIs with params, which was missing from the previous PR. The e2e sync test has been updated to also test restoration.

**See changes from new commits only here:** [new commits](https://github.com/ray-project/ray/pull/32576/files/f0af5c9c437fb4afc39d248456cb365e76770d0f..90616137c384cf08c1e501f1ebe7b1a050396989)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/impl/tuner_internal.py [hunk 1]
```diff
@@ -399,9 +394,11 @@ def _maybe_sync_down_tuner_state(self, restore_path: str) -> Tuple[bool, str]:
...
         tempdir = Path(tempfile.mkdtemp("tmp_experiment_dir"))
 
-        path = Path(restore_path)
-        download_from_uri(str(path / _TRAINABLE_PKL), str(tempdir / _TRAINABLE_PKL))
-        download_from_uri(str(path / _TUNER_PKL), str(tempdir / _TUNER_PKL))
+        restore_uri = URI(restore_path)
+        download_from_uri(
+            str(restore_uri / _TRAINABLE_PKL), str(tempdir / _TRAINABLE_PKL)
+        )
+        download_from_uri(str(restore_uri / _TUNER_PKL), str(tempdir / _TUNER_PKL))
```
Comment (justinvyu) [inline python/ray/tune/impl/tuner_internal.py:401]: This is new.

File: python/ray/tune/execution/trial_runner.py [hunk 2]
```diff
@@ -292,7 +293,7 @@ def scheduler_alg(self):
     @property
     def _remote_checkpoint_dir(self):
         if self._sync_config.upload_dir and self._experiment_dir_name:
-            return os.path.join(self._sync_config.upload_dir, self._experiment_dir_name)
+            return str(URI(self._sync_config.upload_dir) / self._experiment_dir_name)
         return None
```
Comment (justinvyu) [inline python/ray/tune/execution/trial_runner.py:297]: This is new.

File: python/ray/tune/tests/test_syncer.py [hunk 3]
```diff
@@ -713,6 +732,60 @@ def test_sync_folder_with_many_files_fs(tmpdir):
...
+    )
+    result_grid = tuner.fit()
+
+    assert result_grid.errors
+
+    shutil.rmtree(local_dir)  # Rely on sync-down from cloud
+    tuner = tune.Tuner.restore(
+        str(URI(mock_s3_bucket_uri) / exp_name), resume_errored=True
+    )
+    result_grid = tuner.fit()
```
Comment (justinvyu) [inline python/ray/tune/tests/test_syncer.py:770]: This is new.

File: python/ray/tune/trainable/trainable.py [hunk 4]
```diff
@@ -614,7 +615,7 @@ def _maybe_load_from_cloud(self, checkpoint_path: str) -> bool:
         rel_checkpoint_dir = TrainableUtil.find_rel_checkpoint_dir(
             self.logdir, checkpoint_path
         )
-        external_uri = os.path.join(self.remote_checkpoint_dir, rel_checkpoint_dir)
+        external_uri = str(URI(self.remote_checkpoint_dir) / rel_checkpoint_dir)
```
Comment (justinvyu) [inline python/ray/tune/trainable/trainable.py:618]: This is new.

**Review by krfricke — APPROVED** [review]

**Comment by justinvyu** [comment]
@krfricke Before merging, I want to run `workflows` tests - how do I kick these off?

**Comment by krfricke** [comment]
Ah sorry, that was a quick merge...

usually it would have been `[all_tests]` or changing something in a workflows file.

Well, now it will run on master, let's see if everything works

**Comment by justinvyu** [comment]
Ah ok, hopefully it is ok!

**Comment by xwjiang2010** [comment]
Thanks folks! Will keep an eye out!

**Pull request merged**